sklearn/experimental/tests/test_column_transformer.py

```diff
@@ -0,0 +1,484 @@
+    assert_true(hasattr(ct.transformers_[0][1], 'mean_'))
+
+
+def test_column_transformer_none():
+
+    # one None -> ignore
+    X_array = np.array([[0., 1., 2.], [2., 4., 6.]]).T
+    ct = ColumnTransformer(
+        [('trans1', Trans(), [0]), ('trans2', None, [1])])
+    exp = np.array([[0.], [1.], [2.]])
+    assert_array_equal(ct.fit_transform(X_array), exp)
+    assert_array_equal(ct.fit(X_array).transform(X_array), exp)
+
+    # all None -> return shape 0 array
+    ct = ColumnTransformer(
+        [('trans1', None, [0]), ('trans2', None, [1])])
+    assert_array_equal(ct.fit(X_array).transform(X_array).shape, (3, 0))
+    assert_array_equal(ct.fit_transform(X_array).shape, (3, 0))
+
+
+def test_column_transformer_get_feature_names():
+    X_array = np.array([[0., 1., 2.], [2., 4., 6.]]).T
+    ct = ColumnTransformer([('trans', Trans(), [0, 1])])
+    # raise correct error when not fitted
+    assert_raises(NotFittedError, ct.get_feature_names)
+    # raise correct error when no feature names are available
+    ct.fit(X_array)
+    assert_raise_message(AttributeError,
+                         "Transformer trans (type Trans) does not provide "
+                         "get_feature_names", ct.get_feature_names)
+
+    # working example
+    X = np.array([[{'a': 1, 'b': 2}, {'a': 3, 'b': 4}],
+                  [{'c': 5}, {'c': 6}]], dtype=object).T
+    ct = ColumnTransformer(
+        [('col' + str(i), DictVectorizer(), i) for i in range(2)])
+    ct.fit(X)
+    assert_equal(ct.get_feature_names(), ['col0__a', 'col0__b', 'col1__c'])
+
+
+def test_column_transformer_passthrough():
+    X_array = np.array([[0, 1, 2], [2, 4, 6]]).T
+
+    X_res_first = np.array([0, 1, 2]).reshape(-1, 1)
+    X_res_second = np.array([2, 4, 6]).reshape(-1, 1)
+    X_res_both = X_array
+
+    # default no passthrough
+    ct = ColumnTransformer([('trans', Trans(), [0])])
+    assert_array_equal(ct.fit_transform(X_array), X_res_first)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_first)
+
+    # specify passthrough column
+    ct = ColumnTransformer([('trans1', Trans(), [0])], passthrough=[1])
+    assert_array_equal(ct.fit_transform(X_array), X_res_both)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both)
+
+    # column order is not preserved (passed through added to end)
+    ct = ColumnTransformer([('trans1', Trans(), [1])], passthrough=[0])
+    assert_array_equal(ct.fit_transform(X_array), X_res_both[:, ::-1])
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both[:, ::-1])
+
+    # passthrough='remainder' -> passthrough all none selected columns
+    ct = ColumnTransformer([('trans1', Trans(), [0])], passthrough='remainder')
+    assert_array_equal(ct.fit_transform(X_array), X_res_both)
+    assert_array_equal(ct.fit(X_array).transform(X_array), X_res_both)
+
+    # passthrough when all transformers are None
+    ct = ColumnTransformer([('trans1', None, [0])], passthrough='remainder')
```

Comment:
Hm that is a bit of an edge case, right? We should document that well in the docstring of ``remainder``. One might also expect that ``[0]`` is in it.

Reply:
Yes, this is an edge case, that's the reason I added a test for it :-)

I added some language for it in the docstring explanation of `passthrough`, but it adds quite some text and complexity to the docstring ..